### PR TITLE
fix: destroy puppeteer utility on context destruction

### PR DESCRIPTION
**Pull request by jrandolf-google**

### CODE REVIEW

File: packages/puppeteer-core/src/common/bidi/Sandbox.ts [hunk 1]
```diff
@@ -76,13 +75,9 @@ export class Sandbox implements RealmBase {
   }
 
   async document(): Promise<ElementHandle<Document>> {
-    if (this.#document) {
-      return this.#document;
-    }
-    this.#document = await this.#realm.evaluateHandle(() => {
+    return await this.#realm.evaluateHandle(() => {
```
Comment (Lightning00Blade) [inline packages/puppeteer-core/src/common/bidi/Sandbox.ts:80]: Please add a comment here as well.

File: packages/puppeteer-core/src/common/bidi/Realm.ts [hunk 2]
```diff
@@ -46,6 +46,17 @@ export class Realm extends EventEmitter {
...
     this.#frame = frame;
+
+    // TODO(jrandolf): We should try to find a less brute-force way of doing
+    // this.
+    this.connection.on(
+      Bidi.ChromiumBidi.Script.EventNames.RealmDestroyed,
+      async () => {
+        const util = this.internalPuppeteerUtil;
+        delete this.internalPuppeteerUtil;
+        void (await util)?.dispose().catch(() => {});
```
Comment (Lightning00Blade) [inline packages/puppeteer-core/src/common/bidi/Realm.ts]: We should send it to debugError.

File: packages/puppeteer-core/src/common/bidi/Realm.ts [hunk 3]
```diff
@@ -46,6 +46,17 @@ export class Realm extends EventEmitter {
...
   setFrame(frame: Frame): void {
     this.#frame = frame;
+
+    // TODO(jrandolf): We should try to find a less brute-force way of doing
+    // this.
+    this.connection.on(
+      Bidi.ChromiumBidi.Script.EventNames.RealmDestroyed,
+      async () => {
+        const util = this.internalPuppeteerUtil;
+        delete this.internalPuppeteerUtil;
```
Comment (Lightning00Blade) [inline packages/puppeteer-core/src/common/bidi/Realm.ts]: Just set it to `undefined`

File: packages/puppeteer-core/src/common/bidi/Realm.ts (hunk 2 above)
Comment (jrandolf) [reply packages/puppeteer-core/src/common/bidi/Realm.ts]: We don't want to do this because we assume that it can error here.

Comment (Lightning00Blade) [reply packages/puppeteer-core/src/common/bidi/Realm.ts]: `.dispose().catch(debugError);` why would it be a problem to do this?

Comment (jrandolf) [reply packages/puppeteer-core/src/common/bidi/Realm.ts]: It'd be noisy because we know that the utility can error here.

**Review by Lightning00Blade — APPROVED** [review]

**Pull request merged**